src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs

```diff
@@ -214,6 +217,23 @@ public IEnumerable<PackageRestoreData> GetPackagesRestoreData(string solutionDir
             }
 
             return restoreAuditProperties;
+
+            async Task<HashSet<string>> GetSuppressionsAsync(MSBuildNuGetProject msbuildProject)
```

Comment:
PackageRestoreManager and SolutionRestoreJob have different approaches to getting the suppressions.

PackageRestoreManager switches back and forth for every project, but SolutionRestoreJob switches once. 
Was that intentional?

Reply:
Kind of, but not really. It's more about fitting within the existing software architecture.

SolutionRestoreJob is JTF aware, since the assembly is only used in our VS extension. PackageRestoreManager is in NuGet.PackageManagement, which is ILMerged into nuget.exe, and is not JTF aware. It's also not an abstract class that I could add an abstract method onto and have NuGet.PackageManagement.VisualStudio extend with JTF stuff.

Without a fairly significant refactor (plus new and/or breaking API changes in NuGet.PackageManagement), which honestly I didn't even consider/think about before creating this PR, I don't see how it would be feasible to have PackageRestoreManager avoid spamming VS's UI thread with a lot of small work. We absolutely don't want to restore packages on VS's main thread.

The only other idea which I just came up with, is to copy the pattern that we do for a couple of properties. When we create the VS class that extends NuGetProject, currently it gets a few properties and caches them in the class, so that when NuGet.PackageManagement gets the properties, it doesn't need the UI thread.  We could do the same here?  But it'd still need an API on a type that PackageReference projects implement, so they'd either need to throw InvalidOperationExceptions, or switch to the UI thread for work that no production code path uses.  The fact that NuGet.PackageManagement doesn't have different NuGetProjects for PackageReference and PackagesConfig is harming the good design on this feature.

Reply:
> The fact that NuGet.PackageManagement doesn't have different NuGetProjects for PackageReference and PackagesConfig is harming the good design on this feature.

It does, https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/Projects/BuildIntegratedNuGetProject.cs

Reply:
Well, something weird is going on, because `BuildIntegratedNuGetProject` doesn't extend, `MSBuildNuGetProject`, but `MSBuildNuGetProject` implements `IDependencyGraphProject` and therefore has `GetPackageSpecsAsync` and `GetPackageSpecsAndAdditionalMessagesAsync` methods. Is this due to project.json, despite project.json using xproj and not csproj (excluding UWP)?

Anyway, point remains that the reason SolutionRestoreJob and PackageRestoreManager work a bit differently is due to NuGet's software architecture. Restore seems to go through PackageRestoreManager, and therefore has "bad" performance due to frequent thread switching, twice (once to UI thread, once to thread pool) per project.  In VS, when packages.config restore no-ops, but we want to do the audit check, then we're in SolutionRestoreJob, which is JTF aware, and we can be smarter about switching to the UI thread once.

NuGet.PackageManagement isn't JTF aware, probably because it's used by nuget.exe, so if we want to avoid the thread switching during restore, I think there are two options:

1. add an interface into NuGet.PackageManagement that "abstracts" JTF, and does nothing in outside of VS
2. Make items in MSBuildNugetProject work like properties, where we load them at project load (VsSolutionManager initialization), and then use the cached version without need for the UI thread. I don't think that currently non-SDK style projects have any way to modify NuGetAuditSuppress items without causing a project reload, but if there is, then nuGet will have a bug since it will use stale data.

Unfortunately this is one of those cases where there isn't a clear best solution, at least in my opinion.